File: 필수문항/115_트럭.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int w = Integer.parseInt(st.nextToken());
        int L = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        int[] trucks = new int[n];
        for (int i = 0; i < n; i++) {
            trucks[i] = Integer.parseInt(st.nextToken());
        }

        ArrayList<Integer> lane = new ArrayList<>();
        int weight = 0; // 현재 다리 위 무게
        int now = -1; // 현재 배열에 들어간 트럭 인덱스

        for (int i = 0; i < Math.min(n, w); i++) {
            if (weight + trucks[i] <= L) {
                weight += trucks[i];
                lane.add(i);
                now = i;
            } else {
                break;
            }
        }

        while (lane.size() < w) {
            lane.add(-1);
        }

        if (now == n - 1) {
            System.out.println(w + now + 1);
        } else {
            weight = 0;
            int lp = 0;
            int rp = w - 1;

            for (int i = 0; i <= rp; i++) {
                if (lane.get(i) != -1) {
                    weight += trucks[lane.get(i)];
                }
            }

            for (int i = now + 1; i < n; i++) {
                while (true) {
                    if (lane.get(lp) == -1) {
                        if (trucks[i] + weight <= L) {
                            lane.add(i);
                            weight += trucks[i];
                            lp++;
                            rp++;
                            break;
                        } else {
                            lane.add(-1);
                            lp++;
                            rp++;
                        }
                    } else {
                        if (weight - trucks[lane.get(lp)] + trucks[i] <= L) {
                            lane.add(i);
                            weight += -trucks[lane.get(lp)] + trucks[i];
                            lp++;
                            rp++;
                            break;
                        } else {
                            lane.add(-1);
                            weight -= trucks[lane.get(lp)];
                            lp++;
                            rp++;
                        }
                    }
                }
            }
            System.out.println(lane.size() + w);
        }
    }
}

File: 필수문항/115_트럭.py
import sys

input = sys.stdin.readline

# 트럭수, 다리길이, 최대하중
n, w, L = map(int, input().split())
trucks = list(map(int, input().split()))

lane = []
weight = 0  # 현재 다리 위 무게
now = -1  # 현재 배열에 들어간 트럭 인덱스
for i in range(min(n, w)):
    if weight + trucks[i] <= L:
        weight += trucks[i]
        lane.append(i)
        now = i
    else:
        break

while len(lane) < w:
    lane.append(-1)

if now == n - 1:
    print(w + now + 1)
else:
    weight = 0
    lp, rp = 0, w - 1
    for i in range(rp + 1):
        if lane[i] != -1:
            weight += trucks[lane[i]]

    for i in range(now + 1, n):
        while True:
            if lane[lp] == -1:
                if trucks[i] + weight <= L:
                    lane.append(i)
                    weight += trucks[i]
                    lp += 1
                    rp += 1
                    break
                else:
                    lane.append(-1)
                    lp += 1
                    rp += 1
            else:
                if weight - trucks[lane[lp]] + trucks[i] <= L:
                    lane.append(i)
                    weight += -trucks[lane[lp]] + trucks[i]
                    lp += 1
                    rp += 1
                    break
                else:
                    lane.append(-1)
                    weight -= trucks[lane[lp]]
                    lp += 1
                    rp += 1
    print(len(lane) + w)
